---
layout: post
title: Promise
categories: Java
description: Java8新特性Promise
keywords: Promise
typora-root-url: ..
---

# Promise

**定义**：代表了未来将要发生的事件，用来传递异步操作的消息

## **2个特点：**

1. 对象的状态不受外界影响
   1. Promise 对象代表一个异步操作，有3种状态：
      1. pending: 初始状态，不是成功或失败状态。
      2. fulfilled: 意味着操作成功完成。
      3. rejected: 意味着操作失败。
   2. 只有异步操作的结果，可以决定当前是哪一种状态，任何其他操作都无法改变这个状态
2. 一旦状态改变，就不会再变，任何时候都可以得到这个结果
   1. 状态改变的2种可能：
      1. 从 Pending 变为 Resolved
      2. 从 Pending 变为 Rejected
   2. 只要这两种情况发生，状态就凝固了，不会再变了，会一直保持这个结果

## 优缺点

### 优点

- 免了层层嵌套的回调函数
- 提供统一的接口，使得控制异步操作更加容易

### 缺点

- 无法取消 Promise，一旦新建它就会立即执行，无法中途取消
- 如果不设置回调函数，Promise 内部抛出的错误，不会反应到外部
- 当处于 Pending 状态时，无法得知目前进展到哪一个阶段

## 示例

### 创建

```
const enterNumber = () => {

  return new Promise((resolve, reject) => {

    const userNumber = Number(window.prompt("Enter a number (1 - 3):"));

    const randomNumber = Math.floor(Math.random() * 3 + 1);
 

    if(userNumber === randomNumber) {

      resolve('You guessed right!')

    } else {

      reject(new Error(`Wrong! The correct number was ${randomNumber}`))

    }

  });

};
```

```
const main = () => {

  enterNumber().then((message) => {

    console.log(message);

  }).catch((error) => {

    console.log(error.message);

  })

}


or

const main = async () =>{
	try{
		const message = await enterNumber()
		console.log(message)
	}catch(error){
		console.log(error)
	}
}

main()
```

## promise中如何取到[[PromiseResult]]

当返回值validate 为一个[Promise](https://so.csdn.net/so/search?q=Promise&spm=1001.2101.3001.7020)对象时

letvalidate=that.selectValidEvent();

console.log( validate);	